src/runtime_src/tools/xclbinutil/SectionSmartNic.cxx

```diff
@@ -0,0 +1,318 @@
+                      const std::string& fromRelativeDir,
+                      std::vector<char>& buffer) {
+  // Build the path to our file of interest
+  boost::filesystem::path filePath(fromRelativeDir);
+  filePath /= fileName;
+
+  // Open the file
+  std::ifstream file;
+  file.open(filePath.string(), std::ifstream::in | std::ifstream::binary);
+  if (!file.is_open())
+    throw std::runtime_error("ERROR: Unable to open the file for reading: " + fileName);
+
+  // Get the file size
+  file.seekg(0, file.end);
+  std::streamsize fileSize = file.tellg();
+  file.seekg(0, file.beg);
+
+  // Resize the buffer and read in the array
+  buffer.resize(fileSize);
+  file.read(buffer.data(), fileSize);
+
+  // Make sure that the entire buffer was read
+  if (file.gcount() != (std::streamsize)fileSize)
+    throw std::runtime_error("ERROR: Input stream for the binary buffer is smaller then the expected size.");
+
+  file.close();
+}
+
+static
+void
+readAndTransposeByteFiles_recursive(const std::string& scope,
+                                    rapidjson::Value& aValue,
+                                    std::vector<std::string>& byteFileEntries,
+                                    rapidjson::Document::AllocatorType& allocator,
+                                    const std::string& relativeFromDir) {
+  XUtil::TRACE((boost::format("-- Scope: %s") % scope).str());
+
+  // A dictionary
+  if (aValue.IsObject()) {
+    std::vector<std::string> renameCollection;
+    for (rapidjson::Value::MemberIterator itr = aValue.MemberBegin(); itr != aValue.MemberEnd(); ++itr) {
+      std::string currentScope = scope + "::" + itr->name.GetString();
+      // Do we have a match
+      if (std::find(byteFileEntries.begin(), byteFileEntries.end(), currentScope) != byteFileEntries.end()) {
+        std::string myString = itr->name.GetString();
+        renameCollection.push_back(myString);
+      }
+
+      readAndTransposeByteFiles_recursive(currentScope, itr->value, byteFileEntries, allocator, relativeFromDir);
+    }
+
+    // Now that we are out of that "nasty" Iterator loop, rename the updated keys.
+    // TODO: Refactor code to have the new file name be data driven
+    for (const std::string& oldKey : renameCollection) {
+      const static std::string removeSubName = "_file";
+      std::string newKey = oldKey;
+      auto index = newKey.find(removeSubName);
+      if (index == std::string::npos)
+        continue;
+      newKey.replace(index, removeSubName.size(), "");
+
+      // Now rename keys by swapping their values
+      rapidjson::Value::MemberIterator itrOldKey = aValue.FindMember(oldKey.c_str());
+      aValue.AddMember(rapidjson::Value(newKey.c_str(), allocator).Move(), itrOldKey->value, allocator);
+      aValue.RemoveMember(oldKey.c_str());
+    }
+
+    return;  // -- Return --
+  }
+
+  // An array
+  if (aValue.IsArray()) {
+    for (auto itr = aValue.Begin(); itr != aValue.End(); ++itr) {
```

Comment:
Hmm...are you suggesting that we look for another JSON parser, writer, and schema checker because they don't expose the underlying collection?   If so then let's chat.

Some reference you might find useful in our discussion:
[https://rapidjson.org/classrapidjson_1_1_generic_array.html](https://rapidjson.org/classrapidjson_1_1_generic_array.html)
[https://rapidjson.org/classrapidjson_1_1_generic_object-members.html](https://rapidjson.org/classrapidjson_1_1_generic_object-members.html)